### PR TITLE
feat(deriver): add message to conclusion source indices

### PR DESCRIPTION
**Pull request by vintrocode**
make the deriver indicate which messages a conclusion came from via source ids

## Summary by CodeRabbit

* **New Features**
  * Conclusions now show the message IDs cited as evidence, in citation order.
  * Added filtering for conclusions and documents by cited message ID.
  * Added CLI support for `conclusion list --cites <message_id>`.
  * Citations can include messages from other peers in the same batch.

* **Documentation**
  * Updated API, SDK, CLI, and MCP documentation with citation fields, filtering options, and legacy/derived conclusion behavior.

### CODE REVIEW

File: src/deriver/prompts.py
Comment (3un01a) [inline src/deriver/prompts.py:1]: This prompt deviates greatly from the original prompt.
Needs to be tested against the current Neuromancer model to see if any negative effects occur.

Maybe we should keep this prompt out from this PR and scope this PR only to the batch message.

Comment (matthewlanders) [reply src/deriver/prompts.py:1]: @vintrocode I know you did some testing. Good to remove the prompt from this PR or is it required for the other changes?

**Review by VVoruganti — COMMENTED** [review]
So main thing to note here is that the `batch_message_ids` that were added as a new field in this draft would include also the IDs from the interleaved messages from other speakers. 

This means that an explicit conclusion for peer A can reference a message made by peer B in the conversation. Historically we have been authorative in saying a conclusion can only be made for a peer based on something the Peer has actually said, otherwise any claim by a separate peer could lead to nonsensical conclusions being made

File: src/deriver/prompts.py
Comment (matthewlanders) [reply src/deriver/prompts.py:1]: Note that if we do revert most of the prompt changes, we will still need to update the prompt to give the model context about `source_indices `.

**Comment by matthewlanders** [comment]
> So main thing to note here is that the `batch_message_ids` that were added as a new field in this draft would include also the IDs from the interleaved messages from other speakers.
> 
> This means that an explicit conclusion for peer A can reference a message made by peer B in the conversation. Historically we have been authorative in saying a conclusion can only be made for a peer based on something the Peer has actually said, otherwise any claim by a separate peer could lead to nonsensical conclusions being made

A few things important to clarify:

1.) The context from which conclusions are derived hasn't changed — the deriver has always seen other speakers' messages as context ("Extract ALL observations from the target peer's messages, using others as context" is the pre-existing rule and the prompt is built from every message in the batch, each labeled with its speaker). This PR just makes the model declare which messages it used.

2.) A citation including another peer's message doesn't mean the conclusion rests on their claim. Specifically, the target peer's statements are often uninterpretable alone ("the first one" in response to "do you prefer Python or Rust"). Because the entailment judge only sees cited messages, the interpreting context has to be cited or valid conclusions fail. In both prompt examples the peer's own message is always among the sources.

If anything this makes the "only from what the peer said" stance enforceable for the first time — a conclusion citing zero target-peer messages is now mechanically detectable.

3.) Your comments made me realize that the design of `batch_message_ids` was kind of suboptimal. It was only ever a lookup table for resolving batch-relative indices, and persisting the full batch list per observation was not useful. I just pushed a small change that resolves citations at save time and stores `source_message_ids` — just the IDs of the cited messages. `source_indices` stays in storage as per-derivation debugging provenance only. The model still only ever sees positions, and the peer-filtered `message_ids` field is untouched.

File: src/models.py [hunk 1]
```diff
@@ -472,6 +472,32 @@ def source_ids(self, value: list[str] | None) -> None:
             for i, sid in enumerate(dict.fromkeys(value or []))
         ]
 
+    # Messages the deriver cited as evidence for an explicit conclusion.
+    source_message_links: Mapped[list["DocumentSourceMessage"]] = relationship(
```
Comment (VVoruganti) [inline src/models.py:476]: **[P2] Preserve citation links when backfilling a scope**

The new relationship also needs to be carried through the scope-copy path. In `src/deriver/scope_backfill.py`, `_CopySpec`, `_run_backfill()`, and the destination `Document` construction copy `source_ids` but omit `source_message_ids`. Adding an existing session to a scope therefore loses the citations on every copied explicit conclusion. I reproduced a source conclusion with two cited messages returning `source_message_ids: null` through the API after backfill; filtering the scope's copies by either message also cannot find it. Please carry the cited message IDs through the copy plan and create their links on the destination document.

File: src/crud/document.py [hunk 2]
```diff
@@ -1288,6 +1303,9 @@ def _document_model_from_create(
             session_name=doc.session_name,
             embedding=doc.embedding,
             source_links=build_source_links(doc.source_ids, workspace_name),
+            source_message_links=build_source_message_links(
+                doc.source_message_ids, workspace_name
+            ),
```
Comment (VVoruganti) [inline src/crud/document.py:1308]: **[P2] Merge incoming citations when deduplication reinforces a conclusion**

Citation links are only written when `_document_model_from_create()` inserts a document. The always-on exact-deduplication path in `create_documents()` skips that helper for an existing match, and `_apply_document_row_updates()` only increments `times_derived`. When two messages in the same session independently produce the same conclusion, the second message's citation is silently discarded, even with `deduplicate=False`. I reproduced `times_derived == 2` while filtering by the second message's ID returned no conclusions. Please merge incoming citations into the retained document, preserving order and deduplicating IDs; the within-batch duplicate path needs the same treatment.

**Review by VVoruganti — COMMENTED** [review]
Two citation-preservation issues, both reproduced with database-backed API tests. The existing 203 focused Python tests, 8 CLI tests, and migration test pass.

File: migrations/versions/c4d8e2f6a1b3_add_document_source_messages_table.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+schema = get_schema()
+
+
+def upgrade() -> None:
+    connection = op.get_bind()
+    inspector = sa.inspect(connection)
+
+    if not table_exists("document_source_messages", inspector):
+        op.create_table(
+            "document_source_messages",
```
Comment (VVoruganti) [inline migrations/versions/c4d8e2f6a1b3_add_document_source_messages_table.py:37]: torn on the approach here, but would like to move away from the document language if possible. Would like to rename this to `conclusion_source_messages`

File: src/crud/document.py [hunk 4]
```diff
@@ -53,6 +53,21 @@ def build_source_links(
...
 
 
+def build_source_message_links(
+    message_ids: list[str] | None, workspace_name: str
+) -> list[models.DocumentSourceMessage]:
+    """Convert cited message public ids to DocumentSourceMessage rows, deduped."""
+    if not message_ids:
+        return []
+    return [
+        models.DocumentSourceMessage(
```
Comment (VVoruganti) [inline src/crud/document.py:63]: Same thing if we can name this ConclusionSourceMessage
